### ISSUE TITLE
ElasticSearch Ruby random failed tests

### ISSUE THREAD

**Issue by slavosarik** [travis-ci/travis-ci#3181]
Hello,

I am using private repo https://magnum.travis-ci.com/slavosarik/Crowdex, rails app with elastic service. The issue - there is 2/3 chance, that tests using ElasticSearch will fail. If I am running integration tests on localhost, I dont get this random failed tests. If I am using Travis builds, there are always failed ElasticSearch tests and I need to rerun them 2 or 3 times in row to get build passing. 

This issue persists last month. Always same error for all ElasticSearch tests. 

```
Failure/Error: expect(Experiment.search("Little Experiment", fields: ["title", "desc"])[0].title).to eq "Little Experiment"
     Searchkick::InvalidQueryError:
       [400] {"error":"SearchPhaseExecutionException[Failed to execute phase [query], all shards failed; shardFailures {[Wv4EadvhRIKNOIBjWFshtw][experiments_test][0]: SearchParseException[[experiments_test][0]: from[-1],size[-1]: 
Parse Failure [Failed to parse source [{\"query\":{\"dis_max\":{\"queries\":[{\"match\":{\"title.analyzed\":{\"query\":\"Little Experiment\",\"operator\":\"and\",\"boost\":10,\"analyzer\":\"searchkick_search\"}}},{\"match\":{\"title.analyzed\":{\"query\":\"Little Experiment\",\"operator\":\"and\",\"boost\":10,\"analyzer\":\"searchkick_search2\"}}},{\"match\":{\"title.analyzed\":{\"query\":\"Little Experiment\",\"operator\":\"and\",\"boost\":1,\"fuzziness\":1,\"max_expansions\":3,\"analyzer\":\"searchkick_search\"}}},{\"match\":{\"title.analyzed\":{\"query\":\"Little Experiment\",\"operator\":\"and\",\"boost\":1,\"fuzziness\":1,\"max_expansions\":3,\"analyzer\":\"searchkick_search2\"}}},{\"match\":{\"desc.analyzed\":{\"query\":\"Little Experiment\",\"operator\":\"and\",\"boost\":10,\"analyzer\":\"searchkick_search\"}}},{\"match\":{\"desc.analyzed\":{\"query\":\"Little Experiment\",\"operator\":\"and\",\"boost\":10,\"analyzer\":\"searchkick_search2\"}}},{\"match\":{\"desc.analyzed\":{\"query\":\"Little Experiment\",\"operator\":\"and\",\"boost\":1,\"fuzziness\":1,\"max_expansions\":3,\"analyzer\":\"searchkick_search\"}}},{\"match\":{\"desc.analyzed\":{\"query\":\"Little Experiment\",\"operator\":\"and\",\"boost\":1,\"fuzziness\":1,\"max_expansions\":3,\"analyzer\":\"searchkick_search2\"}}}]}},\"size\":100000,\"from\":0,\"fields\":[]}]]]; nested: QueryParsingException[[experiments_test] [match] analyzer [searchkick_search] not found]; }{[Wv4EadvhRIKNOIBjWFshtw][experiments_test][1]: SearchParseException[[experiments_test][1]: from[-1],size[-1]: Parse Failure [Failed to parse source [{\"query\":{\"dis_max\":{\"queries\":[{\"match\":{\"title.analyzed\":{\"query\":\"Little Experiment\",\"operator\":\"and\",\"boost\":10,\"analyzer\":\"searchkick_search\"}}},{\"match\":{\"title.analyzed\":{\"query\":\"Little Experiment\",\"operator\":\"and\",\"boost\":10,\"analyzer\":\"searchkick_search2\"}}},{\"match\":{\"title.analyzed\":{\"query\":\"Little Experiment\",\"operator\":\"and\",\"boost\":1,\"fuzziness\":1,\"max_expansions\":3,\"analyzer\":\"searchkick_search\"}}},{\"match\":{\"title.analyzed\":{\"query\":\"Little Experiment\",\"operator\":\"and\",\"boost\":1,\"fuzziness\":1,\"max_expansions\":3,\"analyzer\":\"searchkick_search2\"}}},{\"match\":{\"desc.analyzed\":{\"query\":\"Little Experiment\",\"operator\":\"and\",\"boost\":10,\"analyzer\":\"searchkick_search\"}}},{\"match\":{\"desc.analyzed\":{\"query\":\"Little Experiment\",\"operator\":\"and\",\"boost\":10,\"analyzer\":\"searchkick_search2\"}}},{\"match\":{\"desc.analyzed\":{\"query\":\"Little Experiment\",\"operator\":\"and\",\"boost\":1,\"fuzziness\":1,\"max_expansions\":3,\"analyzer\":\"searchkick_search\"}}},{\"match\":{\"desc.analyzed\":{\"query\":\"Little Experiment\",\"operator\":\"and\",\"boost\":1,\"fuzziness\":1,\"max_expansions\":3,\"analyzer\":\"searchkick_search2\"}}}]}},\"size\":100000,\"from\":0,\"fields\":[]}]]]; nested: QueryParsingException[[experiments_test] [match] analyzer [searchkick_search] not found]; }{[Wv4EadvhRIKNOIBjWFshtw][experiments_test][2]: SearchParseException[[experiments_test][2]: from[-1],size[-1]: Parse Failure [Failed to parse source [{\"query\":{\"dis_max\":{\"queries\":[{\"match\":{\"title.analyzed\":{\"query\":\"Little Experiment\",\"operator\":\"and\",\"boost\":10,\"analyzer\":\"searchkick_search\"}}},{\"match\":{\"title.analyzed\":{\"query\":\"Little Experiment\",\"operator\":\"and\",\"boost\":10,\"analyzer\":\"searchkick_search2\"}}},{\"match\":{\"title.analyzed\":{\"query\":\"Little Experiment\",\"operator\":\"and\",\"boost\":1,\"fuzziness\":1,\"max_expansions\":3,\"analyzer\":\"searchkick_search\"}}},{\"match\":{\"title.analyzed\":{\"query\":\"Little Experiment\",\"operator\":\"and\",\"boost\":1,\"fuzziness\":1,\"max_expansions\":3,\"analyzer\":\"searchkick_search2\"}}},{\"match\":{\"desc.analyzed\":{\"query\":\"Little Experiment\",\"operator\":\"and\",\"boost\":10,\"analyzer\":\"searchkick_search\"}}},{\"match\":{\"desc.analyzed\":{\"query\":\"Little Experiment\",\"operator\":\"and\",\"boost\":10,\"analyzer\":\"searchkick_search2\"}}},{\"match\":{\"desc.analyzed\":{\"query\":\"Little Experiment\",\"operator\":\"and\",\"boost\":1,\"fuzziness\":1,\"max_expansions\":3,\"analyzer\":\"searchkick_search\"}}},{\"match\":{\"desc.analyzed\":{\"query\":\"Little Experiment\",\"operator\":\"and\",\"boost\":1,\"fuzziness\":1,\"max_expansions\":3,\"analyzer\":\"searchkick_search2\"}}}]}},\"size\":100000,\"from\":0,\"fields\":[]}]]]; nested: QueryParsingException[[experiments_test] [match] analyzer [searchkick_search] not found]; }{[Wv4EadvhRIKNOIBjWFshtw][experiments_test][3]: SearchParseException[[experiments_test][3]: from[-1],size[-1]: Parse Failure [Failed to parse source [{\"query\":{\"dis_max\":{\"queries\":[{\"match\":{\"title.analyzed\":{\"query\":\"Little Experiment\",\"operator\":\"and\",\"boost\":10,\"analyzer\":\"searchkick_search\"}}},{\"match\":{\"title.analyzed\":{\"query\":\"Little Experiment\",\"operator\":\"and\",\"boost\":10,\"analyzer\":\"searchkick_search2\"}}},{\"match\":{\"title.analyzed\":{\"query\":\"Little Experiment\",\"operator\":\"and\",\"boost\":1,\"fuzziness\":1,\"max_expansions\":3,\"analyzer\":\"searchkick_search\"}}},{\"match\":{\"title.analyzed\":{\"query\":\"Little Experiment\",\"operator\":\"and\",\"boost\":1,\"fuzziness\":1,\"max_expansions\":3,\"analyzer\":\"searchkick_search2\"}}},{\"match\":{\"desc.analyzed\":{\"query\":\"Little Experiment\",\"operator\":\"and\",\"boost\":10,\"analyzer\":\"searchkick_search\"}}},{\"match\":{\"desc.analyzed\":{\"query\":\"Little Experiment\",\"operator\":\"and\",\"boost\":10,\"analyzer\":\"searchkick_search2\"}}},{\"match\":{\"desc.analyzed\":{\"query\":\"Little Experiment\",\"operator\":\"and\",\"boost\":1,\"fuzziness\":1,\"max_expansions\":3,\"analyzer\":\"searchkick_search\"}}},{\"match\":{\"desc.analyzed\":{\"query\":\"Little Experiment\",\"operator\":\"and\",\"boost\":1,\"fuzziness\":1,\"max_expansions\":3,\"analyzer\":\"searchkick_search2\"}}}]}},\"size\":100000,\"from\":0,\"fields\":[]}]]]; nested: QueryParsingException[[experiments_test] [match] analyzer [searchkick_search] not found]; }{[Wv4EadvhRIKNOIBjWFshtw][experiments_test][4]: SearchParseException[[experiments_test][4]: from[-1],size[-1]: Parse Failure [Failed to parse source [{\"query\":{\"dis_max\":{\"queries\":[{\"match\":{\"title.analyzed\":{\"query\":\"Little Experiment\",\"operator\":\"and\",\"boost\":10,\"analyzer\":\"searchkick_search\"}}},{\"match\":{\"title.analyzed\":{\"query\":\"Little Experiment\",\"operator\":\"and\",\"boost\":10,\"analyzer\":\"searchkick_search2\"}}},{\"match\":{\"title.analyzed\":{\"query\":\"Little Experiment\",\"operator\":\"and\",\"boost\":1,\"fuzziness\":1,\"max_expansions\":3,\"analyzer\":\"searchkick_search\"}}},{\"match\":{\"title.analyzed\":{\"query\":\"Little Experiment\",\"operator\":\"and\",\"boost\":1,\"fuzziness\":1,\"max_expansions\":3,\"analyzer\":\"searchkick_search2\"}}},{\"match\":{\"desc.analyzed\":{\"query\":\"Little Experiment\",\"operator\":\"and\",\"boost\":10,\"analyzer\":\"searchkick_search\"}}},{\"match\":{\"desc.analyzed\":{\"query\":\"Little Experiment\",\"operator\":\"and\",\"boost\":10,\"analyzer\":\"searchkick_search2\"}}},{\"match\":{\"desc.analyzed\":{\"query\":\"Little Experiment\",\"operator\":\"and\",\"boost\":1,\"fuzziness\":1,\"max_expansions\":3,\"analyzer\":\"searchkick_search\"}}},{\"match\":{\"desc.analyzed\":{\"query\":\"Little Experiment\",\"operator\":\"and\",\"boost\":1,\"fuzziness\":1,\"max_expansions\":3,\"analyzer\":\"searchkick_search2\"}}}]}},\"size\":100000,\"from\":0,\"fields\":[]}]]]; nested: QueryParsingException[[experiments_test] [match] analyzer [searchkick_search] not found]; }]","status":400}
```

**Comment by BanzaiMan** [comment]
@slavosarik Sorry to hear about this issue. We can spin up a debug VM for you. Please email support@travis-ci.com.

**Comment by slavosarik** [comment]
Thank you.

**Comment by pleary** [comment]
Any chance you figured out the problem? I'm experiencing a similar problem with Travis, ES and Ruby/Rails and am looking for leads. Thanks!

**Comment by slavosarik** [comment]
I have experimented with build script settings and found out that ES was not working properly due to longer service initialization (starting ES service is async task, so while I were creating db models, ES was not ready yet, so the indices werent created properly). I added to build script

`before_script:
- sleep 20 
`

and since that I have not experienced same issue yet.

**Comment by pleary** [comment]
OK, thanks for the info. I have been using a delay already, and my problems have been appearing in the middle of the test suite, perhaps because the tests are constantly dropping and recreating the indices for a common baseline. We're also using Elasticsearch 2.1 which isn't the Travis default yet. Our .travis.yml is here if you care https://github.com/inaturalist/inaturalist/blob/686335e5de7ee7c36db96cf707ac0551b4dcbbbf/.travis.yml .

Thanks for your quick response! If I find a solution I'll leave another comment.

**Comment by spiderbites** [comment]
I'm having the same issue with tests failing in the middle of the suite, we're also doing lots of dropping and recreating of indices.  Did you find any solution?

**Comment by pleary** [comment]
Shoot - I should have responded months ago. I forget exactly what the root cause was, but I know that one thing I added was a method to make sure the index exists after recreating before continuing with tests (https://github.com/inaturalist/inaturalist/commit/4d3c639cab48b60e2868b048ad5dd1b30d4fca39#diff-93830fa29d616f7c87903d08b5b1b29aR146). I think what was happening was we were creating the indexes all the time, but occasionally ES was just a little too slow and we'd try to query the index in a test but ES hadn't caught up yet.

Another possibility was it was just user error and I was running a query that was trying to sort on a field that I failed to include in the default mapping.

Good luck!
